---
title: "Complete QTL and Candidate Search Tutorial"
author: "John T. Lovell"
date: "`r Sys.Date()`"
output: 
  pdf_document:
    toc: true
    toc_depth: 1
    number_sections: true
    fig_caption: yes
---

***

  [email: johntlovell@gmail.com](johntlovell@gmail.com)  --  [website: lovelleeb.weebly.com](lovelleeb.weebly.com)  --  [github: github.com/jtlovell/qtlTools](https://github.com/jtlovell/qtlTools)

***

```{r, include=FALSE, message=FALSE}
suppressPackageStartupMessages(library(knitr))
knitr::opts_chunk$set(echo = TRUE, 
                      message=FALSE, 
                      error=FALSE, 
                      warning = FALSE, 
                      fig.width = 5.5, fig.height = 3.5)

suppressPackageStartupMessages(library(qtlTools))
suppressPackageStartupMessages(library(qtlDesign))
suppressPackageStartupMessages(library(ggplot2))
```





# Setup
For this tutorial, we need a couple datasets and R packages. 

1. The [`R/qtl`](www.rqtl.org) R pacakge. 
  - `install.packages("qtl")`
  - `library(qtl)`
  
2. The [`qtlTools`](https://github.com/jtlovell/qtlTools) R package, in development. 
  - If you don't have the devtools package, run:
    - `install.packages("devtools")`
  - `devtools::install_github("jtlovell/qtlTools")`
  - `library(qtlTools)`
  
3. The [`ggplot2`](http://ggplot2.org/) R package
  - `install.pacakges("ggplot2")`
  - `library(ggplot2)`
  
4. The [tutorial dataset]() which contains an R image with the following files:
  - A `cross` object, which stores the genetic map, genotype and phenotype data needed to do QTL mapping
  - A `gff` file with the physical positions of all genes
  - A `vcf_summary` file with the counts of various types of mutations in each gene
  - `parentExp` and `parentInfo` files with gene expression and experimental design data for the parental genotypes
  - A `f2Exp` file with gene expression for the F2 population 
  - A `map` file that contains the physical AND mapping positions of all markers. 
  

**To get the tutorial dataset directly from R, run the following:**
```{r}
download.file(paste0("https://github.com/jtlovell/Rqtl_tutorials/blob/master/", # website
                     "completeQTL_tutorial_data.rda", # data file name
                     "?raw=true"), # use the 'raw' data
              destfile = "~/Desktop/completeQTL_tutorial_data.rda", 
              method="wget", quiet = TRUE)
load("~/Desktop/completeQTL_tutorial_data.rda")
```


**Here are the files names:**
```{r}
ls()
```


**The R/qtl cross object summary (`cr`):**
```{r}
summary(cr)
```


**The genetic and physical positions of markers (`map`):**
```{r}
knitr::kable(head(map))
```


**The genome feature file annotation (`gff`):**
```{r}
knitr::kable(head(gff))
```


**The parental expression (`parentExp`) and experimental design (`info`) data:**
```{r}
knitr::kable(parentExp[1:5,1:5])
knitr::kable(head(parentInfo))
```


**It's important to note that these datasets match ...**
```{r}
identical(rownames(parentInfo), colnames(parentExp))
```


**The gene-by-gene summary of non-synonymous DNA variants (`vcf_summary`):**
```{r}
knitr::kable(head(vcf_summary))
```


**The expression dataset for the mapping population (`f2Exp`):**
```{r}
knitr::kable(f2Exp[1:5,1:5])
```


**It's important the the line IDs in the cross object match the row names of the f2Exp dataset**
```{r}
identical(rownames(f2Exp), as.character(getid(cr)))
```




# Overview - developing expectations for QTL interval sizes and number of candidate genes.
One goal of QTL mapping is to determine the genes that have the potential to affect phenotypic trait variation ... candidate genes.
While highly accurate, the precision of linkage mapping in experimental populations is limitted by the scale of recombination - QTL often span 100's of kb's and contain many gene models.

To illustrate this, let's model the size of confidence intervals for a single large effect QTL in a 200 individual F2 population.

```{r, echo = F, fig.cap = "Distribution of QTL confidence intervals (calculated as the 1-LOD drop interval) from 100 simulated linkage maps. For each map, a single QTL was simulated with a LOD score of approximately 25 (Additive effect = 1, dominance effect = 0)"}
set.seed(42)
qtlWidth<-sapply(1:100,function(x){
  diff(
    lodint(
      scanone(
        sim.cross(
          map = sim.map(len = 100, n.mar = 50, include.x = F),
          model = rbind(c(1,45,1,0)),
          n.ind = 200,
          type = "f2"),
        method = "hk"))[c(1,3),2])
})
hist(qtlWidth, breaks = 20, xlab = "Width of QTL confidence interval (cM)",
     main = "Histogram of 100 QTL intervals")
```

Note that most QTL's are 5-20cM wide. A typical *A. thaliana* genetic map is 500cM and 212Mb ... ~400Kb / cM. If the ~30k genes are distributed evenly, we might expect 300-1200 genes under any given QTL with a LOD of 25. As such, there is considerable work remaining after finding a QTL peak.

**Note** *Higher LOD QTL, more recombination (e.g. RIL/MAGIC populations) or more individuals will reduce the width of QTL intervals. If there is little additional information in a mapping population, it is best to get information from as many recombinant genotypes as possible (e.g. subsequent fine mapping or very large populations).*

\newpage

# The phenotype data
Here we will look at an F2 population of Panicum hallii - a genetic model for C4 grasses. Importantly, there are two environmental treatments (Dry and Wet)... these are covariates, which are stored as `covar` and included in all analyses.

```{r}
covar<-data.frame(
  covar = as.numeric( # turn treatment classification into 1/2's
    as.factor( # turn character "wet/dry" into a factor
      pull.pheno(cr, pheno.col = "Treatment")))) # pull the phenotype data from cr
```

```{r, fig.cap = "Distribution of phenotypic variation. Data is skewed, but this isn't really a problem if we are looking at a single QTL ... We use permutations to test significance, which account for non-normal distributions."}
ggplot(pull.pheno(cr), 
       aes(x = phenotype, fill = Treatment))+
  scale_fill_manual(values = c("darkred","cornflowerblue"))+
  geom_histogram(binwidth = .15)+
  theme_jtlbar()
```

# Run a 1-way QTL scan
This is the first step in QTL mapping ... for each marker / pseudo marker in the genome, we test the hypothesis H0: no QTL vs. HA: 1 QTL.
Since here we also have the added complexity of an experimental covariate, our explicit likelihood ratio test is:
H0: y ~ 1 ... HA: y ~ QTL + Treatment + QTL*Treatment

```{r,echo = T}
s1<-scanone(cr,
            method = "hk", # use the haley-knott regression
            pheno.col = "phenotype",
            intcovar = covar, # force treatment to interact with the QTL
            addcovar = covar) # for there to be an additive effect of treatment
```

To test for significance, we must run permutations... 
```{r,echo = T}
s1p<-scanone(cr,
            n.perm = 100,
            method = "hk", # use the haley-knott regression
            pheno.col = "phenotype",
            intcovar = covar, # force treatment to interact with the QTL
            addcovar = covar, # for there to be an additive effect of treatment
            verbose = F)
```

```{r, echo = T, fig.cap = "The distribution of maximum LOD scores from scanone permutations (100). The 90th%ile of the permutations is presented by the vertical "}
plot(s1p)
abline(v = quantile(s1p,.95), col = "red", lty=2)
```

To see if we have a QTL, we plot the one-way scan results along with the permutation threshold
```{r,echo = T, fig.cap = "One-way QTL scan for loci associated with phenotypic variation."}
plot(s1, ylab = "LOD")
add.threshold(out = s1, perms = s1p, col = "red", lty=2)
```

# Finding a QTL peak and confidence interval

We now want to determine the width of the QTL interval. To do this, we calculate the LOD profile. In this case, since we only have one QTL, the LOD profile will be identical to the one-way QTL scan.

```{r, echo = T}
# make a QTL model
q<-makeqtl(cr, 
           chr = max(s1)$chr, 
           pos = max(s1)$pos, 
           what = "prob")
# generate a LOD profile
r<-refineqtl(cross = cr,
             qtl = q, # the qtl model
             formula = y~Q1+covar+Q1*covar, # the formula, w/ covariate interactions
             method = "hk", # haley-knott regression
             covar = covar, # covariate dataset
             pheno.col = "phenotype",
             verbos = F)
```

Using the LOD profile, determine the QTL confidence interval as the 90% approximate Bayesian credible interval.
```{r}
cis<-calcCis(cross = cr,
             mod = r, # the qtl model with calculated LOD profile
             pheno.col = "phenotype",
             lodint = F, # calculate bayes confidence intervals
             prob = .9 # use 90% probability
             )
```

Finally, plot the LOD profile and add the confidence interval around the QTL peak.
```{r, echo = T, fig.cap = "Chromosome 3 LOD profile including the position of the QTL peak and confidence interval (the red point and line segment)."}
plotLodProfile(r,
               xlab = "Chr03 mapping position",
               ylab = "LOD")
segmentsOnPeaks(cr,
                calcCisOutput = cis,
                chr = 3,
                mod = r,
                int.y = 8,
                col = rgb(1,0,0,.5),
                lwd = 3, pt.cex = 1)
```

\newpage

# Finding genes under a QTL
The first step towards finding candidate genes is finding out which gene models lie within the interval. To do this, we need two datasets:

  - An annotation (e.g. gff3) file with the location of all genes
  - The mapping and physical position of markers

**Using these datasets, we can determine the mapping postion of all genes**

This can be calculated using any number of approaches. Here we employ a marker-by-marker linear model. In short, we pull two adjacent markers and calculate the slope and intercept of the relationship between physical (bp) and genetic mapping (cM) positions. We then pull all genes that have physical positions between those markers and predict their mapping from physical positions.
```{r, echo = T}
geneCM<-findGenecM(cross = cr, # the cross object
                   marker.info = map, # the genetic map with 
                   # a column including bp position
                   gff = gff[gff[,3]=="gene",], # the gff file. 
                   #For speed I cull to only 'genes'
                   attributeParse = "Name=", 
                   # the strings to drop from the gene name
                   seqnameParse = "Chr0"
                   # the strings to drop from the Chr identifier
                   # so that they match the chromosome names in the cross.
                   )
```

```{r, fig.height = 5, fig.width = 5, fig.cap = "The relationship between phyical and mapping positions of all genes. The red points are the markers in the genetic map. The black points are all genes in the annotation. Note the low recombination rates in the pericentromeric regions ... not surprising given that this is a grass."}
kable(head(geneCM[,-c(2,6,7,8)]))

ggplot(map, aes(x = bp/1000000, y = pos))+
  geom_point(col = "red", pch = 8, cex = .5)+
  geom_point(data = geneCM, pch = ".")+
  facet_wrap(~chr)+theme_jtl()+
  labs(x = "physical position (Mbp)",
       y = "genetic mapping position (cM)")
```

\newpage
## Pull out genes in the intervals
```{r, echo = T}
candGenes<-findGenesInterval(findGenecM.output = geneCM,
                             calcCis.output = cis)
```

```{r}
print(candGenes)
```
**Our list of candidate genes is now 162**

\newpage
# Culling candidate gene list by expression of the parents
To definitively clone a QTL, we need to get the list of candidates down to a managable number ... this number depends on how easy it is to get and assay transgenics in your system.
The first step towards narrowing down the candidate gene list is to determine what molecular mechanism you think could potentially be causing phenotypic variation. Here, we have data suggesting that the phenotype is driven by transcript abundance of a regulatory gene. 
Therefore, we expect the parents to have differential expression of the causal gene in the QTL interval. This means that any genes with NO differential expression between the parents can be disregarded in our candidate gene search.

Here, we test differential expression using the `DESeq2` R package and a pipeline in the `deTools` development version R package

```{r}
devtools::install_github("jtlovell/deTools")
suppressPackageStartupMessages(library(deTools))

# if you don't have DESeq2 installed ... run this:
# source("https://bioconductor.org/biocLite.R")
# biocLite("DESeq2")
suppressPackageStartupMessages(library(DESeq2))
```

```{r}
de.parent<-pipeDESeq2(counts = parentExp, info = parentInfo,
           formula = " ~ type + Treatment + type*Treatment",
           reduced = " ~ Treatment",
           testNames = "geno")
de.test<-de.parent$LRT_results
de.test$signif<-ifelse(de.test$padj_geno<=0.01, "sig","NS")
de.candidates<-de.test$gene[de.test$signif == "sig"]
print(de.candidates)
```
**We now have 62 candidates**

# Culling candidates by DNA sequence variation
We may also expect phenotypes to be driven by *non-synonymous variants* ... In this case, one may generate a vcf (variant call format) matrix that details all sequence variants between the parents of the linkage mapping population, then annotate this file using snpeff and count the number of non-synonymous variants for each gene. Those genes with >0 non-synonymous variants could be your culled candidate gene list.

Here, we have annotated the genes in the interval and qualified the effects of variants as `LOW` (synonymous), `MODERATE` (non-synonymous or similar), `HIGH` (frameshift, premature stop, etc.).

```{r}
## Need another packages here ... reshape2
vcf_toplot<-reshape2::melt(vcf_summary[,c("geneID","HIGH","MODERATE")])

nmodhigh<-with(vcf_toplot, tapply(value, geneID, sum))
nmodhigh<-nmodhigh[order(nmodhigh)]
vcf_toplot$geneID<-factor(vcf_toplot$geneID, levels = names(nmodhigh))
ggplot(vcf_toplot, aes(y = value, x = geneID, fill = variable))+ 
  geom_bar(stat = "identity")+
  theme_jtlbar()+
  theme(axis.text.x = element_blank())+
  labs(x = "Candidate Gene", y = "n MODERATE & HIGH variants", fill = "SNP Effect")
```

So, we need to set some threshold for a potential candidate gene ... lets just arbitrarily say that we need >3 MODERATE effect OR >0 HIGH effect variant
```{r}
seq_candidates<-vcf_summary$geneID[vcf_summary$HIGH>=1 | vcf_summary$MODERATE>3]
```

How many of the differnetial expression candidates are also DNA sequence candidates?
```{r}
table(de.candidates %in% seq_candidates)

de_and_seq_candidates<-de.candidates[de.candidates %in% seq_candidates]
print(de_and_seq_candidates)
```
**Our list of candidates is down to 48**


# Culling by expression of the mapping population.
If we believe that a QTL is caused by a gene expression polymorphism, then we are explicitly only interested in candidate genes that have *cis*-eQTL. That is, a local variant drives expression of the candidate gene.

To find this, we load normalized expression data from all genes in the confidence interval. And treat these expression values as phenotypes

```{r}
cr2<-cr
cr2$pheno<-f2Exp

allCands<-colnames(f2Exp)

s1.cand<-scanone(cr2, 
                 pheno.col = allCands, # all the genes in the region 
                 method = "hk",
                 addcovar = covar, 
                 intcovar = covar, 
                 chr = 3) # just look at Chr3 (speeds things up)
s1.cand.perm <- scanone(cr2, 
                        pheno.col = allCands, 
                        method = "hk",
                        addcovar = covar, 
                        intcovar = covar, 
                        chr = 3, 
                        n.perm = 100, verbose =F)
```

Now lets look and see which of these candidate genes have a good cis-eQTL in the region
```{r, fig.cap = "scanone plots for all of the genes that are found in the qtl interval on Chr3"}
thresholds = summary(s1.cand.perm, alpha = .05) # calculate permutation thresholds
max.s1s<-apply(s1.cand[,-c(1:2)], 2, max) # calculate QTL peak LOD
max.pos<-s1.cand$pos[apply(s1.cand[,-c(1:2)], 2, which.max)] # find the position of the QTL Peak
plot(max.pos,max.s1s,
     xlab = "Mapping position of QTL peak",
     ylab = "LOD score of QTL peak")
rect(xleft = cis$lowposition, xright = cis$highposition,
     ybottom = min(thresholds), ytop = max(max.s1s),
     col = rgb(1,0,0,.2))
```

Now, find all genes in the QTL interval with a significant QTL in the interval. 
```{r}
cisEqtl_candidates<-allCands[which(max.s1s>=thresholds &
        max.pos>=cis$lowposition &
        max.pos <= cis$highposition)]
table(de_and_seq_candidates %in% cisEqtl_candidates)
seqExpCis_candidates<-de_and_seq_candidates[de_and_seq_candidates %in% cisEqtl_candidates]
print(seqExpCis_candidates)
```

** Our list of candidate genes is down to 24**

# Rank Candidate genes.
Here we use the covariance of expression and trait of interest in mapping population (Lovell et al. 2015, Plant Cell) to rank the potential effects of a candidate gene.
To do so, we iteratively add, then remove the gene expression of a candidate to the QTL model. Those that reduce the signal of the focal QTL are better candidates than those which do not.

```{r}
cand.mat<-pull.pheno(cr2,seqExpCis_candidates)
```

```{r, fig.cap = "Output from the candidate gene search. The red line indicates the intial LOD profile. Each grey line represents the LOD profile when that gene is added to the model as a covariate. Lower values at the QTL peak are indicative of potentially greater effect on the focal phenotype."}
covs<-covScanQTL(cross = cr,
                 pheno.col = "phenotype",
                 qtl = q, # the qtl model from the initial scan
                 expression.covariates = cand.mat, # matrix of gene expression data
                 qtl.method = "hk",
                 nperm = 20, # how many permutation to do to test for significance?
                 addcovar = covar,
                 intcovar = covar,
                 ylab = "LOD")
```

Here are the top 6 candidates:
```{r}
kable(head(covs))
```

And all significant candidates:
```{r}
covs$candidateID[covs$perm.p<=0.05]
```

# Conclusions
So, thats really it. The next step is to look at these candidates, figure out which ones might make sense ... then do the molecular biology to acutally nail down causation.
GOOD LUCK!
